File: core/src/main/scala/com/klibisz/elastiknn/storage/BitBuffer.scala
package com.klibisz.elastiknn.storage

/**
  * Minimal abstraction to simplify storing a series of bits as a single scalar value, convertible to a byte array.
  */
sealed trait BitBuffer {
  def putOne(): Unit
  def putZero(): Unit
  def toByteArray: Array[Byte]
}

object BitBuffer {

  class IntBuffer(barr: Array[Byte] = Array.empty) extends BitBuffer {
    private var buf = 0
    private var i = 0
    override def putOne(): Unit = {
      buf += IntBuffer.powers(i)
      i += 1
    }
    override def putZero(): Unit = i += 1
    override def toByteArray: Array[Byte] = barr ++ UnsafeSerialization.writeInt(buf)
  }

  object IntBuffer {
    private val powers = (0 until 32).map(1 << _).toArray
  }

}

File: elastiknn-core/src/main/java/com/klibisz/elastiknn/storage/BitBuffer.java
package com.klibisz.elastiknn.storage;

public interface BitBuffer {
    void putOne();
    void putZero();
    byte[] toByteArray();

    class IntBuffer implements BitBuffer {

        private final byte[] prefix;
        private int i = 0;
        private int b = 0;

        public IntBuffer(byte[] prefix) {
            this.prefix = prefix;
        }

        public IntBuffer() {
            this.prefix = new byte[0];
        }

        @Override
        public void putOne() {
            this.b += (1 << this.i);
            this.i += 1;
        }

        @Override
        public void putZero() {
            this.i += 1;
        }

        @Override
        public byte[] toByteArray() {
            byte[] barr = UnsafeSerialization.writeInt(b);
            byte[] res = new byte[prefix.length + barr.length];
            System.arraycopy(prefix, 0, res, 0, prefix.length);
            System.arraycopy(barr, 0, res, prefix.length, barr.length);
            return res;
        }
    }

}